Refactor target packages to use interfaces

swernli opened the pull request:
This change refactors the target packages to define interfaces for the natively supported Q# callables, which provides an important layer of indirection that lets the Q#/C# in the target package call into those interfaces while the simulator implements those interfaces, all without either taking a direct reference on the other. This means I can once again unify the simulator into a single package, where it dynamically detects the target package intrinsics to override at runtime via reflection.

Fixes #432 

Major changes in this package:

1. Introduce Simulation/TargetDefinition/Interfaces that defines the package with the union of all interfaces for all target packages and target specific interfaces that collect the set of gates supported by that target (expected to eventually be auto-generated from Q# definitions).

2. Implementations of the callables in Simulation/TargetDefinition/Intrinsic that implement those abstract classes and call into the interfaces (also expected to be auto-generated, perhaps replacing the abstract classes altogether).

3. Adapted the full state QuantumSimulator, ToffoliSimulator, and ResourceEstimator to use the interfaces such that the single simulator package can now be used with all three target packages (simulator no longer has compile time dependency on the target package). This means Simlators.Impl and Simulators could be combined back into a single folder, and Simulators.Type1 and Simulators.Type2 eliminated.

4. Updated test project definitions to include references to both the simulator and the desired target package, matching the pattern we expect callers to use.

5. Moved the QuantumProcessor implementation into the QsharpCore package so that this implementation can remain the same. Converting it to the interface based approach would be a breaking change, but to keep it the same required putting it in  a place that could safely compile against both QsharpCore and the SimulatorBase class. For the moment that means that the QsharpCore target package has a compile time dependency on the Simulators package, which doesn't match the design intent but doesn't cause any problems. Eventually, QuantumProcessor can either be moved to its own package or removed entirely in favor of broader adoption of the interface based design.

6. Updates to C# generation that provide an opt-in mechanism for intrinsic Q# callables to be generated as concrete classes instead of abstract, and the generated code calls into the interface classes defined as part of 1 (thanks @ScottCarda-MS). This means the intrinsic callables that are part of the target package no longer need to define a child class that registers as an override of the abstract parents.

swernli commented:
A note on point 6 from above: @samarsha had an idea for how the supported constructor parameters of the custom simulator provided to the unit test or EntryPoint can be dynamically detected so that we can avoid the reflection-based search for the TargetIntrinsics type. I'm going to try that approach, but I wanted to start getting feedback on the design and changes now while I continued to refine that part of the changes.

bettinaheim commented on src/Simulation/Simulators/QCTraceSimulator/QCTraceSimulator.Primitive.Measure.cs:
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Why is this file new? There seems to be a change to how Measure works for the tracer. Can you elaborate a bit on that?

swernli replied:
Yes, previously QCTraceSimulator defined an override for `M` instead of `Measure` and relied on the behavior that any C# implementation would automatically override the Q# even if it wasn't body intrinsic (`M` has always been defined as a decomposition on `Measure` but the simulators would unconditionally override that). Now that the built-in simulators cannot inherit from and directly override the callabes from the target package, this simulator needs to actually participate in the decomposition+intrinsic infrastructure, meaning it needs to provide an implementation for `Measure` instead.

bettinaheim commented on src/Simulation/Simulators/QuantumSimulator/ApplyControlledX.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public partial class QuantumSimulator
+    {
+        public virtual Func<(Qubit, Qubit), QVoid> ApplyControlledX_Body() => (args) =>
```
Thinking about how the generated code needs to look like, since this is what would be implemented manually, we should expose it with a more natural signature and push the "plumbing" into the generated piece.
```
public virtual void ApplyControlledX_Body(Qubit control, Qubit target) { ... }
```

swernli replied:
I agree, that is much nicer! I'm going to rework the interfaces to use this approach, but first I'd like to get the e2e build working to ensure I have a clean foundation for further changes.

swernli replied:
Ok, I've adjusted the interface style to match this suggestion. Now I'll work on changing the C# generation so that we don't need to use the type parameters for overriding.

swernli replied:
Thanks @ScottCarda-MS for making the changes needed to support concreate class generation for intrinsics! It makes all of this cleaner.

bettinaheim commented on src/Simulation/Simulators/QuantumSimulator/NativeImports.cs:
```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Much nicer to have them all here!

bettinaheim commented on src/Simulation/Simulators/ToffoliSimulator/ToffoliSimulator.cs:
```diff
@@ -23,7 +24,7 @@ public enum ToffoliDumpFormat
     /// The Toffoli simulator implementation class.
     /// </summary>
-    public partial class ToffoliSimulator : SimulatorBase
+    public partial class ToffoliSimulator : SimulatorBase, IQsharpCore
```
What work is still missing for the ToffoliSimulator and the Tracer/ResourcesEstimator to process code compiled for a all three gate sets? Aside from that support, is there any other difference between how the three simulators behave/what they support?

swernli replied:
All that is missing is the need to implement the native gates for those packages, which include things like `ApplyUncontrolledX` and `SWAP`. This isn't hard, I just didn't think it was necessary for the current PR. I agree that we should consider supporting all targets on these simulators, but there are some open questions, namely:

- How do we want to refactor the ResourcesEstimator, and will that change the approach and how we support multiple targets?

- Does it make sense to represent ToffoliSimulator as it's own target package, since it defines a gate set, rather than allow someone to use a specific gate set and then target Toffoli which doesn't actually support that set?

bettinaheim commented on src/Simulation/TargetDefinitions/Interfaces/IGate_ApplyControlledX.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public interface IGate_ApplyControlledX : IOperationFactory
+    {
+        Func<(Qubit, Qubit), QVoid> ApplyControlledX_Body();
```
Before we ship these interfaces, it would be good to do the change to make the exposed methods in a more user friendly format (see my previous comment).

swernli replied:
Totally agreed. I'll make sure to adjust that as part of this PR.

samarsha commented on Simulation.sln:
```diff
@@ -69,12 +69,10 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{BC562DAE-FE2
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Simulation", "Simulation", "{9008B252-2DF7-404B-B626-D4497BB70A05}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Quantum.QSharp.Foundation", "src\Simulation\QsharpFoundation\Microsoft.Quantum.QSharp.Foundation.csproj", "{DB45AD73-4D91-43F3-85CC-C63614A96FB0}"
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Quantum.QSharp.Foundation", "src\Simulation\QsharpFoundation\Microsoft.Quantum.Qsharp.Foundation.csproj", "{DB45AD73-4D91-43F3-85CC-C63614A96FB0}"
```
For comparison, it seems that C# and F# are both consistently written with a capital S (CSharp and FSharp).

samarsha replied:
> a lot less invasive than changing all Sdk properties

[MSBuild property names are case-insensitive.](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild?view=vs-2019#BKMK_Items)

swernli replied:
While MSBuild doesn't care, Linux and dotnet core do, which is why this worked fine for local build and failed e2e. If there is a casing difference, Linux's case-sensitive file system will say the project reference is not found, and fail compilation with missing dependencies.

I agree with Bettina that we should determine a consistent pattern. For now I'm just going to try and avoid casing changes for existing files and ensure my new files compile, and then we can revisit this its own effort. I filed #449 to track this.

samarsha replied:
Yes, I didn't mean that we shouldn't use consistent casing - I meant that I don't think changing the MSBuild properties to use capital S (currently they're written Qsharp) would be an invasive change, since it seems like MSBuild ignores case differences.

swernli replied:
Ah, I see what you mean. Yes, that makes sense!

swernli commented on AdvantageBenchmark/privateBuild/Program.cs:
```diff
@@ -24,7 +24,7 @@ public static void Main(string[] args)
             if (args.Length > 2) loopCnt = Convert.ToInt32(args[2]);
 
-            using (var sim = new QuantumSimulator()) 
+            using (var sim = new QuantumSimulator(typeof(Microsoft.Quantum.Intrinsic.TargetIntrinsics))) 
```
Not of all simulators, as some simulators may decide to provide their own overrides instead of needing to find the interface shims. I have some ideas about how to remove the need for this, which I'll expound on in the other comments.

swernli commented on src/Simulation/Simulators.Tests/CoreTests.cs:
```diff
@@ -33,12 +33,12 @@ public void BasicExecution()
-            Assert.Null(ex);
-            Assert.Equal(1, exitCode);
-            Assert.Contains("NotImplementedException", error.ToString());
+            // ProcessRunner.Run("dotnet", exe, out var _, out StringBuilder error, out int exitCode, out Exception ex);
```
Whoops! I don't think that was intentional, that is a leftover from some testing. I'll uncomment them and see if they pass now, and properly address them if they do not.

swernli commented on src/Simulation/Simulators.Tests/QuantumSimulatorTests/BasicTests.cs:
```diff
@@ -290,17 +290,17 @@ public void TestSimpleGateCheckQubits()
-                    typeof(Intrinsic.H),
-                    typeof(Intrinsic.S),
-                    typeof(Intrinsic.T),
+                    // typeof(Intrinsic.H),
```
Likewise, this was a merge mistake from some experiments I was doing. They should be uncommented.

swernli commented on src/Simulation/CsharpGeneration/EntryPoint.fs:
```diff
@@ -152,12 +152,15 @@ let private entryPointClass context entryPoint =
         property "Info" (sprintf "EntryPointInfo<%s, %s>" argTypeName returnTypeName)
                         (ident callableName <|.|> ident "Info")
+    let targetIntrinsicProperty =
+        property "TargetIntrinsicsType" "Type" (invoke (ident "typeof") ``(`` [(ident "Microsoft.Quantum.Intrinsic.TargetIntrinsics")] ``)``)
```
No matter which target package is being used, it always defines that type, which has the concrete classes used as overrides for the abstract classes generated for body intrinsic callables. However, this is awkward, and thinking about it more, I realized that if I go ahead and update the C# generation to have an option for producing non-abstract classes that call into the interface directly, then the overrides will no longer be needed and I can remove this type parameter and related infrastructure. Since that's the direction we want to go in anyway and it will allow removal of these awkward workarounds, I think it's worth doing as part of this change.

swernli replied:
With the interface changes worked on by @ScottCarda-MS all of this infrastructure now goes away, and there is no longer a need to infer the originating type for intrinsic overrides :)

swernli commented on src/Simulation/Simulators/QCTraceSimulator/QCTraceSimulator.cs:
```diff
@@ -398,13 +398,19 @@ public string[] GetMetricStatisticNames(string metric)
         /// the <paramref name="config"/> parameter.
         /// </summary>
-        public QCTraceSimulator(QCTraceSimulatorConfiguration config) : base(config) { }
+        public QCTraceSimulator(QCTraceSimulatorConfiguration config) : base(FindTargetIntrinsicsType(), config) { }
```
This would also go away if the C# generation created concrete classes instead of abstract classes, so I'm going to take a shot at removing it in favor of that approach.

swernli commented:
@bettinaheim This branch now has all the latest updates from @ScottCarda-MS and myself, so the PR is ready for re-review. I'll work on fixing the e2e build once this goes into the feature branch and I do another merge from master.

bettinaheim reviewed (commented):
I haven't quite made it through the PR yet, but let me post the pending comments

swernli commented on src/Simulation/CsharpGeneration/SimulationCode.fs:
```diff
@@ -69,16 +78,24 @@ module SimulationCode =
         let hasMultipleDefinitions() = if context.byName.ContainsKey op.Name then context.byName.[op.Name].Length > 1 else false
         let sameNamespace = match context.current with | None -> false | Some n -> n.Namespace = op.Namespace
 
+        let namespaces = if isConcreteIntrinsic context then autoNamespacesWithInterfaces else autoNamespaces
```
As written, this might have that effect, but I don't think it should be required. Generating the concrete intrinsics with calls into the interfaces and generating the interfaces themselves can be done at the same time while following the same set of assumptions. As long as all the files exist before the C# is compiled then it will be fine; the concrete intrinsics and the interfaces don't actually need any information other than the Q# declarations and don't depend on any generated details from each other.

swernli commented on src/Simulation/QsharpCore/Microsoft.Quantum.QSharp.Core.csproj:
```diff
@@ -1,9 +1,13 @@
 
   <PropertyGroup>
     <Description>Core support for the Q# programming language.</Description>
+    <GenerateConcreteIntrinsic>true</GenerateConcreteIntrinsic>
```
Actually, that name would imply something other than what is happening here. These changes do not include automatic generation of the interfaces, those are still written by hand (see response blow about needing to check in those manual files). This does include the optional flag that lets a Q# project opt into the interface design such that the `body intrinsic` Q# declarations will produce concrete classes that call into the interfaces when generating C# (instead of abstract classes that must be overridden by an inheritor). This behavior has to be opt-in because otherwise it would be a breaking change. This way, any external developers writing their own `body intrinsic` operations can continue to use the old method while new projects like the target packages can use the new infrastructure.

swernli commented on src/Simulation/Simulators.Tests/Circuits/VerifyUnitary.qs:
```diff
@@ -55,7 +55,7 @@ namespace Microsoft.Quantum.Simulation.Simulators.Tests.Circuits {
             // Now test control... We'll have 3 control qubits.
             // We will run the test with 1..3 controls at a time.
-            let ctrlsCount = 3;
+            let ctrlsCount = 2;//3;
```
Whoops! That is an oversight, left over from testing. I need to put that back and confirm that everything still works as expected. Thanks for catching that!

swernli commented on src/Simulation/Simulators/QuantumSimulator/ApplyControlledX.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    public partial class QuantumSimulator
+    {
+        public virtual void ApplyControlledX_Body(Qubit control, Qubit target)
```
Right, I remember now that you had mentioned this before. I can update these files and the concrete intrinsic generation to assume double underscore.

swernli commented on src/Simulation/TargetDefinitions/Interfaces/IGate_ApplyControlledX.cs:
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
As mentioned in a response above, these files are not yet autogenerated, so I do need to check them in. The concrete intrinsics that call into these interfaces ARE autogenerated, which was necessary to remove the inheritance/override patterns, but generating these is just a nice to have that won't change any other infrastructure, so I thought we could leave that to a later update.

bettinaheim replied:
Ok, it would be nice to not have to check in everything that we autogenerate, but I agree that we can have them as normal code files for now.

swernli commented on src/Simulation/TargetDefinitions/Interfaces/Microsoft.Quantum.Targets.Interfaces.csproj:
```diff
@@ -0,0 +1,30 @@
+    <PlatformTarget>x64</PlatformTarget>
+  </PropertyGroup>
+
+  <PropertyGroup>
```
Good call. I'll do another pass to find common properties that can be pulled into either an existing props file or a new one for this layer of projects.

swernli commented:
@bettinaheim I believe I've addressed the comments you have so far, thanks for looking at this! I know you mentioned you still wanted to review a little more, so I'd be happy to go over some of the files together if you have other questions/feedback you'd like to discuss!
